web3/contract.py

```diff
@@ -589,6 +591,21 @@ def __getattr__(self, function_name):
     #
     # Private Helpers
     #
+    _return_data_normalizers = []
+
+    @classmethod
+    def _normalize_return_data(cls, output_types, output_data):
+        data_tree = abi_data_tree(output_types, output_data)
+
+        for normalizer in cls._return_data_normalizers:
+            data_tree = data_tree_map(normalizer, data_tree)
```

Comment:
to remove mutation...

```python
from cytoolz import pipe
data_tree = pipe(
    output_data, 
    partial(abi_tree_data, output_types),
    *(partial(data_tree_map, normalizer) for normalizer in cls._return_data_normalizers)
)
```

Readability could be cleaned up by decomposing this into multiple lines with intermediate variables.

Reply:
Ok, I took another swing at it in the functional style. I'm roughly familiar with the principles, but it doesn't come naturally to me yet. LMK what you think.

Reply:
(updated in #312)